File: Inboxes/Node.js/Cli/tests/inboxes.test.js
const {
  sendMessage,
  listMessages,
  readMessage,
  deleteMessage,
} = require("../inboxes");

let fresh;

beforeEach(() => {
  jest.resetModules();
  fresh = require("../inboxes");
});

describe("Inbox CLI", () => {
  test("sendMessage creates message", () => {
    const msg = fresh.sendMessage("alice", "Hello", "World");
    expect(msg.id).toBe(1);
    expect(msg.from).toBe("alice");
    expect(msg.subject).toBe("Hello");
    expect(msg.body).toBe("World");
    expect(msg.read).toBe(false);
    expect(msg.created_at).toBeDefined();
  });

  test("listMessages returns all", () => {
    fresh.sendMessage("alice", "S1", "B1");
    fresh.sendMessage("bob", "S2", "B2");
    expect(fresh.listMessages()).toHaveLength(2);
  });

  test("readMessage marks as read", () => {
    fresh.sendMessage("alice", "Test", "Body");
    const msg = fresh.readMessage(1);
    expect(msg).not.toBeNull();
    expect(msg.read).toBe(true);
    const msg2 = fresh.readMessage(1);
    expect(msg2.read).toBe(true);
  });

  test("deleteMessage removes it", () => {
    fresh.sendMessage("alice", "Del", "Me");
    expect(fresh.listMessages()).toHaveLength(1);
    expect(fresh.deleteMessage(1)).toBe(true);
    expect(fresh.listMessages()).toHaveLength(0);
  });

  test("listAfterDelete shows remaining", () => {
    fresh.sendMessage("alice", "Keep", "Me");
    fresh.sendMessage("bob", "Delete", "This");
    fresh.deleteMessage(2);
    const msgs = fresh.listMessages();
    expect(msgs).toHaveLength(1);
    expect(msgs[0].id).toBe(1);
  });

  test("readMessage nonexistent returns null", () => {
    expect(fresh.readMessage(999)).toBeNull();
  });

  test("deleteMessage nonexistent returns false", () => {
    expect(fresh.deleteMessage(999)).toBe(false);
  });
});

File: Inboxes/Elixir/Cli/test/inboxes_test.exs
defmodule InboxesTest do
  use ExUnit.Case, async: true

  alias Inboxes

  test "send_message creates an unread message" do
    assert {:ok, m} = Inboxes.send_message([], "Alice", "Hello", "Hi there")
    assert m.id == 1
    assert m.from == "Alice"
    assert m.subject == "Hello"
    assert m.body == "Hi there"
    assert m.read == false
    assert m.created_at != nil
  end

  test "send_message increments ids" do
    {:ok, m1} = Inboxes.send_message([], "Alice", "Hello", "Hi")
    {:ok, m2} = Inboxes.send_message([m1], "Bob", "World", "Hey")
    assert m1.id == 1
    assert m2.id == 2
  end

  test "list_messages returns all" do
    {:ok, m1} = Inboxes.send_message([], "Alice", "Hello", "Hi")
    {:ok, m2} = Inboxes.send_message([m1], "Bob", "World", "Hey")
    assert Inboxes.list_messages([m1, m2]) == [m1, m2]
  end

  test "read_message marks as read" do
    {:ok, m} = Inboxes.send_message([], "Alice", "Hello", "Hi")
    assert {:ok, read} = Inboxes.read_message([m], 1)
    assert read.read == true
  end

  test "read_message missing returns not_found" do
    assert Inboxes.read_message([], 99) == {:error, :not_found}
  end

  test "delete_message removes by id" do
    {:ok, m} = Inboxes.send_message([], "Alice", "Hello", "Hi")
    assert {:ok, ^m} = Inboxes.delete_message([m], 1)
  end

  test "delete_message missing returns not_found" do
    assert Inboxes.delete_message([], 99) == {:error, :not_found}
  end
end